devtools/src/arcs-selector.js

```diff
@@ -0,0 +1,171 @@
+        cursor: pointer;
+        line-height: 20px;
+      }
+      [count] {
+        color: white;
+        background-color: var(--dark-gray);
+        padding: 2px 6px;
+        border-radius: 20px;
+      }
+      [name] {
+        margin-left: 6px;
+      }
+      [name][none] {
+        font-style: italic;
+        color: var(--dark-gray);
+      }
+      .triangle {
+        margin-left: 12px;
+      }
+      iron-dropdown {
+        box-shadow: var(--drop-shadow);
+        background-color: white;
+        padding: 8px 0;
+        min-width: 160px;
+      }
+      [list] {
+        /* Overrides attributes set by the iron-dropdown */
+        max-width: 50vw !important;
+        max-height: 50vh !important;
+      }
+      [entry] {
+        line-height: 20px;
+        padding: 4px 12px;
+        cursor: pointer;
+      }
+      [entry][active] {
+        font-weight: bold;
+      }
+      [entry]:hover {
+        color: #fff;
+        background-color: var(--highlight-blue);
+      }
+      [annotation] {
+        font-style: italic;
+      }
+    </style>
+    <div selector on-click="_openDropdown"><span count>[[arcs.length]]</span><span name none$="[[!active]]">[[_activeName(active)]]</span><span class="triangle devtools-small-icon" expanded></span></div>
+    <iron-dropdown id="dropdown" horizontal-align="left" horizontal-offset="-6" vertical-align="top" vertical-offset="23">
+      <div slot="dropdown-content" list>
+        <template is="dom-repeat" items="[[arcs]]">
+          <div entry active$=[[item.active]] on-click="_arcSelected">[[item.name]] <span annotation>[[item.annotation]]</span></div>
+        </template>
+      </div>
+    </iron-dropdown>`;
+  }
+  static get is() { return 'arcs-selector'; }
+
+  static get properties() {
+    return {
+      arcId: {
+        type: String,
+        reflectToAttribute: true,
+        computed: '_extractId(active)',
+        notify: true,
+      },
+    };
+  }
+
+  constructor() {
+    super();
+    this.active = null;
+    this.arcs = [];
+    this.messages = new Map();
+  }
+
+  onRawMessageBundle(messages) {
+    const messagesToForward = [];
+    for (const msg of messages) {
+      switch (msg.messageType) {
+        case 'page-refresh':
+          this.arcs = [];
+          this.active = null;
+          this.messages.clear();
+          this.$.dropdown.close();
+          this.emitFilteredMessages([
+            {messageType: 'reset'},
+            {messageType: 'hard-reset'},
+          ]);
+          return;
+        case 'arc-available': {
+          const id = msg.meta.arcId;
+          const name = id.substring(id.indexOf(':') + 1);
+          const item = {id, name};
+          if (name.endsWith('-null')) item.annotation = '(Planning)';
+          if (!this.messages.has(id)) this.messages.set(id, []);
+          this.push('arcs', item);
+          this._select(item);
+          break;
+        }
+        case 'arc-transition':
+          Promise.resolve().then(() => {
```

Comment:
Why is the promise needed?

Reply:
Thank for keeping me honest. Probably not needed, removed. The thinking was that we first go though all messages, and then select an arc, but that's in principle not needed.